File: src-tauri/src/events.rs
use serde::Serialize;

use crate::models::{JudgeVerdict, PipelineStage, StageStatus};

/// Emitted when the pipeline begins execution.
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineStartedPayload {
    pub run_id: String,
    pub prompt: String,
    pub workspace_path: String,
}

/// Emitted when a pipeline stage transitions status.
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineStagePayload {
    pub run_id: String,
    pub stage: PipelineStage,
    pub status: StageStatus,
    pub iteration: u32,
}

/// Emitted for each line of CLI output (stdout or stderr).
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineLogPayload {
    pub run_id: String,
    pub stage: PipelineStage,
    pub line: String,
    pub stream: String,
}

/// Emitted when a notable artefact is produced (diff, review, etc.).
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineArtifactPayload {
    pub run_id: String,
    pub kind: String,
    pub content: String,
    pub iteration: u32,
}

/// Emitted when the pipeline finishes successfully.
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineCompletedPayload {
    pub run_id: String,
    pub verdict: JudgeVerdict,
    pub total_iterations: u32,
    pub duration_ms: u64,
}

/// Emitted when the pipeline encounters an error.
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineErrorPayload {
    pub run_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stage: Option<PipelineStage>,
    pub message: String,
}

/// Emitted when the pipeline pauses to ask the user a question.
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineQuestionPayload {
    pub run_id: String,
    pub question_id: String,
    pub stage: PipelineStage,
    pub iteration: u32,
    pub question_text: String,
    pub agent_output: String,
    pub optional: bool,
}

File: src/types/events.ts
import type { PipelineStage, StageStatus, JudgeVerdict } from "./pipeline";

// ---- Backend event payloads ----

/** Emitted when a new pipeline run begins. */
export interface PipelineStartedEvent {
  runId: string;
  prompt: string;
  workspacePath: string;
}

/** Emitted when a pipeline stage changes status. */
export interface PipelineStageEvent {
  runId: string;
  stage: PipelineStage;
  status: StageStatus;
  iteration: number;
}

/** Emitted for each log line produced by an agent. */
export interface PipelineLogEvent {
  runId: string;
  stage: PipelineStage;
  line: string;
  stream: "stdout" | "stderr";
}

/** Emitted when a pipeline artefact is produced. */
export interface PipelineArtifactEvent {
  runId: string;
  kind: "diff" | "plan" | "plan_audit" | "plan_final" | "review" | "judge" | "executive_summary";
  content: string;
  iteration: number;
}

/** Emitted when the pipeline completes successfully. */
export interface PipelineCompletedEvent {
  runId: string;
  verdict: JudgeVerdict;
  totalIterations: number;
  durationMs: number;
}

/** Emitted when the pipeline encounters an error. */
export interface PipelineErrorEvent {
  runId: string;
  stage?: PipelineStage;
  message: string;
}

/** Pipeline question event — emitted when the pipeline pauses for user input. */
export interface PipelineQuestionEvent {
  runId: string;
  questionId: string;
  stage: PipelineStage;
  iteration: number;
  questionText: string;
  agentOutput: string;
  optional: boolean;
}

/** Answer payload sent back to the backend via the answer_pipeline_question command. */
export interface PipelineAnswer {
  questionId: string;
  answer: string;
  skipped: boolean;
}
